src/linalg.jl

```diff
@@ -33,3 +33,12 @@ for wrapper in [:Adjoint, :Transpose]
     end
   end
 end
+
+function LinearAlgebra.mul!(Y::AbstractVecOrMat, A::AbstractMatrix, B::OneHotLike)
+  _isonehot(B) || return invoke(mul!, Tuple{AbstractArray,AbstractMatrix,AbstractMatrix}, Y, A, B)
+  size(A,2) == size(B,1) || throw(DimensionMismatch("Matrix column must correspond with the OneHot Size $(size(A,2)) ≠ $(size(B,1))")
+)
+  # matmul sometimes wraps in ReshapedArray, taking parent is a simple way to handle that case
+  copyto!(Y, view(A, :, onecold(parent(B))))
```

Comment:
The implementation of `*` uses `NNlib.gather(A, _indices(B))`. It seems like that's the pre-built way of doing `onecold(parent(B)))`, and is perhaps less likely to introduce new edge cases:
```suggestion
  copyto!(Y, view(A, :, _indices(B)))
```

This could also use `NNlib.gather!` instead of `copyto!`. Any comment on this choice? There is some chance that's faster on GPU, but I haven't timed it.

Reply:
Nice call!
```julia
⫏▮▮⫐ @btime CUDA.@sync mul!($Y, $A, $B);  # copyto!
  33.272 μs (286 allocations: 7.72 KiB)
⫏▮▮⫐ @btime CUDA.@sync mul!($Y, $A, $B);  # NNlib.gather!
  14.036 μs (157 allocations: 4.27 KiB)
```
I have it gather except in the mat-vec case where `NNlib.gather!` is not applicable.